### PR TITLE
Async iterator streams wip

### PR DESCRIPTION
**Pull request by H-Plus-Time**
TLDR: It's not pretty, but it works!

Ideally the reader class would open up a long-lived writer, and feed in chunks from the current row group's deserializer -  that _should_ feed directly into arrow-js' RecordBatchStreamReader.

It also turns out it's _easier_ to make a class/object that the ReadableStream constructor accepts than it is to satisfy the AsyncIterator protocol (the need for Symbol.asyncIterator to be present in some capacity requires... unpleasant Reflect calls). Given the widespread support for async iteration on ReadableStreams, it's probably better to just ditch the explicit async iterator bits and stick to streams.

I think the overall surface for this should end up looking like:
```ts
const stream: ReadableStream = read_parquet_async(url);
const reader = await RecordBatchStreamReader.from(stream);
for await (const batch: RecordBatch of reader) {
}
// or just one big table
const combined = await tableFromIPC(stream);
```

and for writing:
```ts
const exportStream: ReadableStream = new Response(combined).stream().pipeThrough(transform_parquet_async());
// do any of the usual things you'd do with a ReadableStream (use as a fetch body, transfer between workers, etc). OR:
const buf = await (new Response(exportStream)).arrayBuffer();
```

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -103,8 +108,17 @@ features = [
...
+  "ReadableStream",
+  "ReadableStreamDefaultController"
 ]
 
+[dependencies.zstd]
+features = [
+  "wasm",
+  "thin"
+]
+default-features = false
```
Comment (kylebarron) [inline Cargo.toml:120]: I've been searching for how to fix this issue for so long! It was the biggest headache in trying to update the repo

Comment (kylebarron) [reply Cargo.toml:120]: I brought this out into its own PR in https://github.com/kylebarron/parquet-wasm/pull/279

File: src/arrow1/writer_properties.rs [hunk 2]
```diff
@@ -22,10 +22,10 @@ impl From<Compression> for parquet::basic::Compression {
         match x {
             Compression::UNCOMPRESSED => parquet::basic::Compression::UNCOMPRESSED,
             Compression::SNAPPY => parquet::basic::Compression::SNAPPY,
-            Compression::GZIP => parquet::basic::Compression::GZIP,
-            Compression::BROTLI => parquet::basic::Compression::BROTLI,
+            Compression::GZIP => parquet::basic::Compression::GZIP(Default::default()),
```
Comment (kylebarron) [inline src/arrow1/writer_properties.rs:25]: ah TIL you can pass `Default::default()` instead of manually using `GzipLevel::default()`

Comment (kylebarron) [reply src/arrow1/writer_properties.rs:25]: Keep this change when you rebase 😄

File: src/arrow1/reader.rs [hunk 3]
```diff
@@ -1,32 +1,31 @@
...
-use parquet::arrow::{ArrowReader, ParquetFileArrowReader};
-use parquet::file::reader::{FileReader, SerializedFileReader};
-use std::sync::Arc;
+use parquet::arrow::arrow_reader::ParquetRecordBatchReaderBuilder;
 
 /// Internal function to read a buffer with Parquet data into a buffer with Arrow IPC Stream data
 /// using the arrow and parquet crates
 pub fn read_parquet(parquet_file: &[u8]) -> Result<Vec<u8>> {
     // Create Parquet reader
     let cursor = Bytes::copy_from_slice(parquet_file);
```
Comment (kylebarron) [inline src/arrow1/reader.rs:11]: unrelated: I realized today looking back at this that this is probably an accidental memory copy of the entire parquet buffer. Taking `Vec<u8>` as the param as the latest `main` does should hopefully improve memory usage here if I'm right

File: src/arrow1/reader_async.rs [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+use crate::log;
+
+use arrow::ipc::writer::StreamWriter;
+use arrow::record_batch::RecordBatchReader;
+use futures::stream::StreamExt;
+use parquet::arrow::async_reader::ParquetRecordBatchStreamBuilder;
+
+use async_compat::CompatExt;
+
+pub async fn read_parquet(url: String) -> Result<Vec<u8>> {
```
Comment (kylebarron) [inline src/arrow1/reader_async.rs:12]: Nice! I've been wanting to have an async api for arrow1 for a while. Especially now that arrow2 is slowing development

File: src/arrow2/reader_async.rs [hunk 5]
```diff
@@ -60,10 +35,14 @@ fn all_elements_equal(arr: &[&Option<String>]) -> bool {
 
 pub async fn read_row_group(
     url: String,
-    // content_length: Option<usize>,
+    content_length: Option<usize>,
     row_group_meta: &RowGroupMetaData,
     arrow_schema: &Schema,
 ) -> Result<Vec<u8>> {
+    let content_length = match content_length {
+        Some(_content_length) => _content_length,
```
Comment (kylebarron) [inline src/arrow2/reader_async.rs:43]: Nit: I think taking off the leading underscore is clearer here

**Comment by kylebarron** [comment]
I bumped a lot of dependencies tonight. Can you merge in the latest `main`?

**Comment by kylebarron** [comment]
My head hurts and so I need to come back to the main gist of this PR. Seems like this async iterator approach might be useful to mention in one of the many wasm-bindgen PRs asking for native async iterator support?

File: .cargo/config.toml [hunk 6]
```diff
@@ -0,0 +1,5 @@
+[build]
+rustflags = ["--cfg=web_sys_unstable_apis"]
```
Comment (kylebarron) [inline .cargo/config.toml:2]: What does this mean? Is this unstable on the Rust side or JS side? Are there browser implications? Does this work on all browsers that support Wasm?

File: src/arrow1/reader_async.rs [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+    // Iterate over record batches, writing them to IPC stream
+    for maybe_record_batch in intermediate {
+        let record_batch = maybe_record_batch?;
+        writer.write(&record_batch)?;
+    }
+    writer.finish()?;
+
+    let writer_buffer = writer.into_inner()?;
+    Ok(writer_buffer.to_vec())
```
Comment (kylebarron) [inline src/arrow1/reader_async.rs:32]: Reading this again I realized that this is probably also a superfluous memory copy (I've learned a lot since I first wrote this!)

In https://github.com/kylebarron/parquet-wasm/pull/280 I changed it to return `output_file` instead of calling `.to_vec`. I'm guessing the latter is likely to have an extra copy

File: tests/js/arrow1.ts [hunk 8]
```diff
@@ -34,6 +34,24 @@ test("read file", async (t) => {
   t.end();
 });
 
+test("read file async", async (t) => {
+  const server = await temporaryServer();
```
Comment (kylebarron) [inline tests/js/arrow1.ts:38]: Amazing; I'd wanted to have an async test for a while, ref https://github.com/kylebarron/parquet-wasm/pull/207

is it possible to test the new stream API too?

File: src/arrow1/reader_async.rs [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+use async_compat::CompatExt;
+
+pub async fn read_parquet(url: String) -> Result<Vec<u8>> {
+    let length = get_content_length(url.clone()).await?;
+    let reader = create_reader(url, length, None);
+    let builder = ParquetRecordBatchStreamBuilder::new(reader.compat()).await?;
+    // quite a few options here - projection masks, row group subselection, etc...
+    let arrow_schema = builder.schema().clone();
+    let parquet_reader = builder.build()?;
+    let intermediate: Vec<_> = parquet_reader.collect().await;
```
[the remark this answers is not in the document]
Comment (H-Plus-Time) [reply src/arrow1/reader_async.rs:19]: Yep, this one does pretty much just do a regular readParquet - we would need to use a similar wrapper struct as in arrow2.

I think for both cases we could mirror some of the internal Builder/Reader structures - with a couple of main usage scenarios:
1. First-K row groups or all row groups - reader.stream(). Consumption can be paused, prematurely stopped without too much hassle (it's all platform-level). The Geoarrow + parquet-wasm observable is what got me thinking about this in the first place (you still want the whole file, but the transform you're applying to it is taxing enough to warrant a stream, which gets you built-in backpressure).
2. Skip a few row groups, pick specific indexes - closer to the current functional approach, you loop over the desired row group indices and dispatch read_row_group calls as usual. Arrow1 may have a significant advantage here, since it looks like it has well-defined filtering specs, at several levels of granularity (down to the PageIndex apparently).

File: src/arrow1/reader_async.rs (hunk 7 above)
Comment (H-Plus-Time) [reply src/arrow1/reader_async.rs:32]: Ah, excellent, direct Vec<u8> returns (I bet things have changed in wasm-bindgen's behaviour, I ended up using a similar thing to copy_vec_to_uint8array a couple of years ago). I'll pluck out the ~3 blocks that still use it.

File: .cargo/config.toml (hunk 6 above)
Comment (H-Plus-Time) [reply .cargo/config.toml:2]: Unstable on the rust side - it's required for use of the `web_sys::ReadableStreamDefaultController` (`ReadableStream` itself doesn't require it). It looks like the controller class isn't globally defined on browsers older than ~Feb 2020 (~7% in most jurisdictions), but calling `new ReadableStream({start, pull, cancel})` works as far back as at least Chrome 57 (Firefox 52-65 and Edge 16-18 appear to be the only versions where WebAssembly support was available before ReadableStream constructors). Eyeballing the caniuse figures, its about 0.1% of browsers with !ReadableStream && Webassembly.

I was a little leery of the potential for the Rust side method names to change, but it looks like the web-sys codegen for optional arguments (<original_method_name>_with_<optional arg>) is pretty stable.

**Comment by H-Plus-Time** [comment]
> 
> I'm not 100% sure how this approach should work with a stream. There _exists_ a [stream interface](https://arrow.apache.org/docs/format/CStreamInterface.html) to the Arrow C Data Interface but I haven't figured out how function pointers work across the Wasm boundary, so I haven't implemented that yet. If we got that working, then we could implement a stream reader on top of the C Stream Interface in JS.
>

Coincidentally, I went down this particular rabbithole around the same time, and reached a somewhat unsatisfying rock-bottom. The C Stream Interface requires the consumer to have access to the same address space as the producer; JS (or JS as a bridge if you were trying to do this across wasm modules) _could_ in principle do this, though there'd need to be at least a few bits on top of the stream interface to make it vaguely safe (something that allocates a blank FFI_ArrowArray and returns a pointer to it, ditto for a schema).

I suppose it really depends on what you want to use the stream for - if it's a transform or aggregation, it makes sense to request the batch be written into the same section of memory (to which the consumer has been given a pointer) and the consumer does whatever with an FFI_ArrowArray view of that memory. A vanilla container that just stashes the current RecordBatch and exposes an FFI view of it would be the best option for JS-land. Shared memory for dynamic-ish linking you'd just call the function pointers as-is (turns out the stream interfaces func pointers are baked in indexes into the wasm module's funcref table); multithreading would require atomics probably.

Probably the more dominant usecase would be a container in rust that incrementally builds up a Vec<FFI_ArrowArray>, where you end up with a long-lived array of FFI_ArrowArray views. The point of the streaming part would be getting a headstart on any expensive post-processing (e.g. tee the stream to kick off copying to WebGL buffers, tee it again for passing through geoarrow area calculations, etc).

**Comment by kylebarron** [comment]
> JS (or JS as a bridge if you were trying to do this across wasm modules) _could_ in principle do this, though there'd need to be at least a few bits on top of the stream interface to make it vaguely safe (something that allocates a blank FFI_ArrowArray and returns a pointer to it, ditto for a schema).

I feel like it's possible, even if you have to watch out for memory safety. At least in the initial implementation, I'd probably require the Rust side to have a function for "provision empty `ArrowArray` struct" that can be passed into the `ArrowArrayStream`'s `get_next`.

Apart from calling the function pointer, everything else is there.

> I suppose it really depends on what you want to use the stream for
>
> Probably the more dominant usecase would be a container in rust that incrementally builds up a Vec<FFI_ArrowArray>

I think the ideal for the stream is to avoid collecting all memory into a vec on the rust side, but rather to have only one parquet chunk at a time in wasm memory.

Maybe it's best to avoid trying to implement the C Stream Interface and instead do our own async iterator approach like you're doing in this PR.

> (turns out the stream interfaces func pointers are baked in indexes into the wasm module's funcref table); multithreading would require atomics probably.

I was never able to figure out how to use the `funcref` table here. If you have experience it would be great to learn!

Last night I wrote down some ideas for a better Arrow-in-Wasm Rust ecosystem. It's not directly related to the topics here but I'll try to put up some sketches soon. (In particular, copying Arrow tables between multiple Wasm memory spaces is super annoying)

File: src/arrow1/ffi.rs [hunk 10]
```diff
@@ -0,0 +1,197 @@
+use std::sync::Arc;
```
Comment (kylebarron) [inline src/arrow1/ffi.rs:1]: Amazing! Thanks for adding this all in

**Comment by H-Plus-Time** [comment]
Update: Ta-dah, a ReadableStream of FFI record batch structs (I've prodded them with the parseField and parseVector functions, identical results to the mainline arrow2 FFI wrappers).

> Probably the more dominant usecase would be a container in rust that incrementally builds up a Vec<FFI_ArrowArray>, where you end up with a long-lived array of FFI_ArrowArray views
>
Turns out I was overly pessimistic about the lifetime of yielded FFIArrowRecordBatches - playing around with the FFIStream implementation (pushed a moment ago), they seem to just stick around as free-floating structs, accumulating them into an array/funneling into a table on the JS side suits that kind of use-case well enough not to bother with anything more explicit.

I've been meaning to add a bit of benchmarking/perf recording, but I can anecdotally say: * You need a ~400Mb/s connection before the context-switching penalty of the stream (JS pull -> Rust iteration -> JS enqueue Uint8Array) gives the advantage to synchronous buffer reads.
* Max memory bump with the IPC approach typically tops out at 50-100MB (e.g. for a multi-GB parquet file, peak memory sits at around that level, assuming you're discarding or shunting the result chunks out to IO).

>
> Maybe it's best to avoid trying to implement the C Stream Interface and instead do our own async iterator approach like you're doing in this PR.
>
Yeah, I had a bit of success with a wrapping struct, but once I realised inter-WASM module calls were a no-go (same module dynamic linking _is_ possible, it's just... a great deal of dark magic), I figured "just leave off the IPC serialization, shove them in FFI wrappers and call it a day".

**Comment by H-Plus-Time** [comment]
>
> I was never able to figure out how to use the funcref table here. If you have experience it would be great to learn!
>
I suspect it's a combination of an maddening default (wasm-bindgen keeps the table non-exported unless you spit out `wasm_bindgen::function_table()` from a public wasm-bindgen'd function), decoy function names (e.g. `table.get(2212)` gets you a `WebAssembly.Function` with an opaque `function 4562() {}` debug presence) and relatively rare fixed-address function pointers (i.e. most of them are dynamically allocated closures, FnMut, etc).

The lightbulb moment was pretty much when I grabbed a view of the FFI_ArrowStream's 40B struct on a few different invocations and noticed the suspiciously not at all pointer-like function pointer addresses:
`Int32 [ 2211, 2212, 2213, 2214, <varying memory address> ]`
A bit of scanning through the .wat representation of the .wasm module confirms those funcrefs to be exactly the right ones(rather than some sort of deterministically generated closure). Call the function returned by `table.get(<funcref id>)` with the right values (thankfully, just a pair of i32 pointers - I ended up picking far-flung as-yet unallocated chunks of memory) and presto, explicit funcref invocation from JS.

TLDR: if a wasm module doesn't opt-in to exporting the table, funcrefs are impossible to use in JS-land (fortunately altering third-party wasm modules is possible).

> Last night I wrote down some ideas for a better Arrow-in-Wasm Rust ecosystem 
>
👀 The parquet, geoparquet, js-polars, geopolars and geoarrow packages are extremely tempting to roll into one giant wasm binary - cracking the module composition problem would be a godsend. I guess shared-everything crates composed at build time might be a way forward (or at least, that's what I gather from the parquet_wasm crate -> geoparquet wasm linkage).

**Comment by kylebarron** [comment]
> Update: Ta-dah, a ReadableStream of FFI record batch structs (I've prodded them with the parseField and parseVector functions, identical results to the mainline arrow2 FFI wrappers - not quite compatible with parseRecordBatch, pretty sure it's that I didn't cast/rejig into StructArrays).

Amazing! As a general note, I think moving to record-batch structs for FFI makes sense because it's less things to move back and forth than always calling `parseField` and `parseVector`, but let's come back to that in another PR.

> Turns out I was overly pessimistic about the lifetime of yielded FFIArrowRecordBatches - playing around with the FFIStream implementation (pushed a moment ago), they seem to just stick around as free-floating structs

On any FFI object, you do have to call `.drop()` when you're done. Otherwise the FFI struct _and_ the array data it points to will not be cleaned up, and you'll run out of memory eventually.

> * You need a ~400Mb/s connection before the context-switching penalty of the stream (JS pull -> Rust iteration -> JS enqueue Uint8Array) gives the advantage to synchronous buffer reads.

Can you clarify what exactly this means? You're saying that the async iterator is worse than a sync buffer read whenever internet is <400mb/s?

I was thinking the main benefit of the async iterator was to be able to start returning arrow batches without waiting for the whole file to finish downloading. 

> * Max memory bump with the IPC approach typically tops out at 50-100MB (e.g. for a multi-GB parquet file, peak memory sits at around that level, assuming you're discarding or shunting the result chunks out to IO).

You're saying that the IPC approach uses 50-100MB more memory than the FFI approach? I would assume this depends on the internal parquet chunk size.

**Comment by kylebarron** [comment]
> wasm-bindgen keeps the table non-exported unless you spit out `wasm_bindgen::function_table()` from a public wasm-bindgen'd function

incredible 🤯 

Ok soon I'm going to explore this a bit more!

> 👀 The parquet, geoparquet, js-polars, geopolars and geoarrow packages are extremely tempting to roll into one giant wasm binary

I'm gonna try to explore some stuff tonight and see how it goes

**Comment by kylebarron** [comment]
In any case I would love to get this PR merged soon! There's a ton of great stuff in here

**Comment by kylebarron** [comment]
I want to push forward the table ideas in https://github.com/kylebarron/arrow-wasm/issues/8 and https://github.com/kylebarron/parquet-wasm/pull/292, but I don't want to make your life harder and get this branch more out of sync with `main`. Is there something I can do to help push this PR over the line?

**Comment by kylebarron** [comment]
I'm also curious if you've looked at/considered https://docs.rs/wasm-streams/latest/wasm_streams/. Is that useful at all here? Or I suppose the stream is a separate approach than an async iterator?

**Comment by H-Plus-Time** [comment]
> 
> I'm also curious if you've looked at/considered https://docs.rs/wasm-streams/latest/wasm_streams/. Is that useful at all here? Or I suppose the stream is a separate approach than an async iterator?
>
I originally had a bit of difficulty using it[^1], but it looks like it does indeed work well enough - there does seem to be a consistent 5% execution time penalty, but I don't think it's worth worrying about (and is probably down to how I'm using it, or a cut corner in the 1st-party approach I've been using). Also, for posterity's sake, compile output size is pretty much the same.

I'll keep this branch around on my fork for reference and start a new branch from main, shouldn't take long.

[^1]: Not Invented Here met _For Science!_. My bad 🫤 .

**Comment by kylebarron** [comment]
I was just curious 🙂 . If you think it's easier/more maintainable to work with `wasm-streams` that sounds great.
